src/index.ts

```diff
@@ -22,16 +26,81 @@ import {
   TransferRequestParams,
 } from './types'
 
+const NETWORK_CHAIN_IDS = {
+  alfajores: 44787,
+  mainnet: 42220,
+}
+
+const fetch = fetchCookie(nodeFetch)
+
 export default class FiatConnectClient implements FiatConectApiClient {
   config: FiatConnectClientConfig
+  _wallet: ethers.Wallet
+  _sessionExpiry?: Date
 
   constructor(config: FiatConnectClientConfig) {
     this.config = config
+    this._wallet = new ethers.Wallet(this.config.privateKey)
+  }
+
+  async _ensureLogin() {
+    const loginResult = await this.login()
+    if (!loginResult.ok) {
+      throw new Error(`Login failed: ${loginResult.val.error}`)
+    }
+  }
+
+  /**
+   * Logs in with the provider and initializes a session.
+   *
+   * @returns a Promise resolving to the literal string 'success' on a
+   * successful login or an Error response.
+   */
+  async login(): Promise<Result<string, ErrorResponse>> {
+    try {
+      if (this._sessionExpiry && this._sessionExpiry < new Date()) {
```

Comment:
This could technically be different from the expiry the server sends back, but that requires complex cookie header parsing, and if there are multiple cookie headers, it would be hard to identify the correct one. So there might be cases where the session actually expires and any request will fail with `Unauthorized`. One potential way of handling this is by having a retry if we get an `Unauthorized` response. Thoughts?

Reply:
So we assume here that the expiration time in the message is the one that's used for the session. I think that's ok

Reply:
Also, I think I probably prefer having the caller dealing with logging (back) in if some call fails due to an `Unauthorized` response (meaning not retrying automatically). The caller might want to implement some custom flow/UI if a call fails due to a stale session vs. creating a brand new session in the first place. Another option would be to have some flag the caller can set to retry automatically if a call fails with `Unauthorized`.

Reply:
Letting the caller handle stale sessions is reasonable, although with the current setup, there's no way for clients to force a login if the expiry times from the client and server differs (either due to clock not in sync or if the server just doesn't use 4h). We could add a `forceNewSession` option to the login method, or just move the condition to `_ensureLogin` and have `login` always create a new session. I like the simplicity of `login` just creating a new session always, but if clients call it multiple times in a short time, it makes a bunch of unnecessary network calls. Maybe that's fine?